addon/corpus.js

```diff
@@ -1,106 +1,231 @@
-            // What can go wrong? Redirects mess up our messaging pipeline.
-            // Modal alerts hang us until the user dismisses them.
-            errorField.appendChild(document.createTextNode(`\nError while freezing ${url}: ${e}`));
+    // Create a new tab with the current url, always as tab[0].
+    // The tabs.onUpdated handler in freezeAllPages() will dispatch a fathom:freeze
+    // event when the tab has completed loading.
+    setCurrentStatus('loading');
+    browser.tabs.create({
+        windowId: windowId,
+        index: 0,
+        url: gUrls[gUrlIndex].url,
+    });
+}
+document.addEventListener('fathom:next', fathomNext, false);
+
+async function fathomFreeze(event) {
+    const windowId = event.detail.windowId;
+    const timer = event.detail.timer;
+    const tab = (await browser.tabs.query({ windowId }))[0];
+
+    setCurrentStatus('freezing');
+    try {
+        // Can't get a return value out of the content script because webpack wraps
+        // our top-level stuff in a function. Instead, we use messaging.
+        await browser.tabs.executeScript(
+            tab.id,
+            { file: '/contentScript.js' }
+        );
+
+        // Call freeze-dry to fetch html.
+        const html = await browser.tabs.sendMessage(
+            tab.id,
+            { type: 'freeze', options: gFreezeOptions }
+        );
+
+        // Clear timeout here so we don't bail out while writing to disk.
+        clearTimeout(timer);
+
+        // Save html to disk.
+        const filename = gUrls[gUrlIndex].filename;
+        await download(html, { filename });
+
+        setCurrentStatus('downloaded');
+    } catch(e) {
+        console.error(tab.url, e.message);
+        // When the tab is closed while things are processing we get errors that
+        // are less than informative.
+        // "Can't access dead object" means the DOM changed while freeze-dry was
+        // running and an object that it's trying to resolve was removed.
+        // "Message manager disconnected" generally means the tab was closed.
+        if (
+            e.message === "can't access dead object" ||
+            e.message === 'Message manager disconnected'
+        ) {
+            setCurrentStatus('error (freezing failed)', true);
+        } else {
+            setCurrentStatus('error (' + e.message + ')', true);
         }
+    } finally {
+        clearTimeout(timer);
+        await browser.tabs.remove(tab.id);
     }
 
-    // Clean up:
-    browser.windows.remove(windowId).catch(() => null);  // Swallow error if window is absent.
-    document.getElementById('freeze').disabled = false;
+    // Done with this url, trigger loading of the next.
+    document.dispatchEvent(new CustomEvent(
+        'fathom:next',
+        { detail: windowId }
+    ));
 }
-document.getElementById('freeze').onclick = freezeAllPages;
+document.addEventListener('fathom:freeze', fathomFreeze, false);
+
+function setCurrentStatus(message, isError=false) {
```

Comment:
Yay, I ♥ statuses! :-) I look forward to taking this even further. Maybe we'll make both it and the Trainer use `<progress>` and look similar. Boy, I wish we could easily decorate the lines in the textarea itself. /me gazes wistfully at the horizon. Anyway, that's a dream for a later PR.